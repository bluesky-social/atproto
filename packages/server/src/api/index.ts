import { createServer } from '../xrpc'
import todoAdx from './todo/adx'
import todoSocial from './todo/social'

export default function () {
  const server = createServer()
  todoAdx(server)
  todoSocial(server)
  return server
}
