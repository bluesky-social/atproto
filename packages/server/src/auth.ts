import * as auth from '@atproto/auth'
import { AuthRequiredError } from '@atproto/xrpc-server'
import * as uint8arrays from 'uint8arrays'
import { DidResolver } from '@atproto/did-resolver'
import express from 'express'
import * as jwt from 'jsonwebtoken'

const BEARER = 'Bearer '
const BASIC = 'Basic '

export type ServerAuthOpts = {
  jwtSecret: string
  didResolver: DidResolver
  adminPass: string
}
export class ServerAuth {
  private _secret: string
  private _adminPass: string
  didResolver: DidResolver
  verifier: auth.Verifier

  constructor(opts: ServerAuthOpts) {
    this._secret = opts.jwtSecret
    this._adminPass = opts.adminPass
    this.didResolver = opts.didResolver
    this.verifier = new auth.Verifier({ didResolver: opts.didResolver })
  }

  createToken(did: string): string {
    return jwt.sign(
      {
        sub: did,
      },
      this._secret,
    )
  }

  getUserDid(req: express.Request): string | null {
    const header = req.headers.authorization || ''
    if (!header.startsWith(BEARER)) return null
    const token = header.slice(BEARER.length)
    const payload = jwt.verify(token, this._secret)
    const sub = payload.sub
    if (typeof sub !== 'string') return null
    if (!sub.startsWith('did:')) return null
    return sub
  }

  getUserDidOrThrow(req: express.Request): string {
    const did = this.getUserDid(req)
    if (did === null) {
      throw new AuthRequiredError()
    }
    return did
  }

  verifyUser(req: express.Request, did: string): boolean {
    const authorized = this.getUserDid(req)
    return authorized === did
  }

  verifyAdmin(req: express.Request): boolean {
    const parsed = parseBasicAuth(req.headers.authorization || '')
    if (!parsed) return false
    const { username, password } = parsed
    if (username !== 'admin') return false
    if (password !== this._adminPass) return false
    return true
  }

  toString(): string {
    return 'Server auth: JWT'
  }
}

export const parseBasicAuth = (
  token: string,
): { username: string; password: string } | null => {
  if (!token.startsWith(BASIC)) return null
  const b64 = token.slice(BASIC.length)
  let parsed: string[]
  try {
    parsed = uint8arrays
      .toString(uint8arrays.fromString(b64, 'base64pad'), 'utf8')
      .split(':')
  } catch (err) {
    return null
  }
  const [username, password] = parsed
  if (!username || !password) return null
  return { username, password }
}

export default ServerAuth
