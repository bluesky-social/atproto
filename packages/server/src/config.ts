import { DidTestRegistry } from './lib/did/did-test'

export interface ServerConfigValues {
  debugMode?: boolean

  scheme: string
  port: number
  hostname: string

  jwtSecret: string

  blockstoreLocation?: string
  databaseLocation?: string

  didTestRegistry?: DidTestRegistry
}

export class ServerConfig {
  constructor(private cfg: ServerConfigValues) {}

  static readEnv() {
    const debugMode = process.env.DEBUG_MODE === '1'

    const hostname = process.env.HOSTNAME || 'localhost'
    let scheme
    if ('TLS' in process.env) {
      scheme = process.env.TLS === '1' ? 'https' : 'http'
    } else {
      scheme = hostname === 'localhost' ? 'http' : 'https'
    }
    const envPort = parseInt(process.env.PORT || '')
    const port = isNaN(envPort) ? 2583 : envPort

    const jwtSecret = process.env.JWT_SECRET || 'jwt_secret'

    const blockstoreLocation = process.env.BLOCKSTORE_LOC
    const databaseLocation = process.env.DATABASE_LOC

    const didTestRegistry = debugMode ? new DidTestRegistry() : undefined

    return new ServerConfig({
      debugMode,
      scheme,
      hostname,
      port,
      jwtSecret,
      blockstoreLocation,
      databaseLocation,
      didTestRegistry,
    })
  }

  get debugMode() {
    return this.cfg.debugMode
  }

  get scheme() {
    return this.cfg.scheme
  }

  get port() {
    return this.cfg.port
  }

  get hostname() {
    return this.cfg.hostname
  }

  get origin() {
    const u = new URL(`${this.scheme}://${this.hostname}:${this.port}`)
    return u.origin
  }

  // @TODO should protect this better
  get jwtSecret() {
    return this.cfg.jwtSecret
  }

  get blockstoreLocation() {
    return this.cfg.blockstoreLocation
  }

  get useMemoryBlockstore() {
    return !this.blockstoreLocation
  }

  get databaseLocation() {
    return this.cfg.databaseLocation
  }

  get useMemoryDatabase() {
    return !this.databaseLocation
  }

  get didTestRegistry() {
    return this.cfg.didTestRegistry
  }
}
