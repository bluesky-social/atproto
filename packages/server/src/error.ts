import { Request, Response } from 'express'
import { logger } from './logger'

export const handler = (err: Error, _req: Request, res: Response) => {
  logger.error({ err }, 'unexpected internal server error')
  res.status(500).send(err.message)
}
