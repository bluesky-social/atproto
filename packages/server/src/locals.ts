import { Response } from 'express'
import pino from 'pino'
import { IpldStore, Repo } from '@atproto/repo'
import { AuthStore, DidableKey } from '@atproto/auth'
import * as plc from '@atproto/plc'
import { Database } from './db'
import { ServerConfig } from './config'
import ServerAuth from './auth'
import { ServerMailer } from './mailer'
import { App } from '.'

export type Locals = {
  logger: pino.Logger
  blockstore: IpldStore
  db: Database
  keypair: DidableKey
  auth: ServerAuth
  config: ServerConfig
  mailer: ServerMailer
}

type HasLocals = App | Response

export const logger = (res: HasLocals): pino.Logger => {
  const logger = res.locals.logger
  if (!logger) {
    throw new Error('No Logger object attached to server')
  }
  return logger as pino.Logger
}

export const blockstore = (res: HasLocals): IpldStore => {
  const blockstore = res.locals.blockstore
  if (!blockstore) {
    throw new Error('No Blockstore object attached to server')
  }
  return blockstore as IpldStore
}

export const db = (res: HasLocals): Database => {
  const db = res.locals.db
  if (!db) {
    throw new Error('No Database object attached to server')
  }
  return db as Database
}

export const keypair = (res: HasLocals): DidableKey => {
  const keypair = res.locals.keypair
  if (!keypair) {
    throw new Error('No Keypair object attached to server')
  }
  return keypair as DidableKey
}

export const config = (res: HasLocals): ServerConfig => {
  const config = res.locals.config
  if (!config) {
    throw new Error('No Config object attached to server')
  }
  return config as ServerConfig
}

export const mailer = (res: HasLocals): ServerMailer => {
  const mailer = res.locals.mailer
  if (!mailer) {
    throw new Error('No Mailer object attached to server')
  }
  return mailer as ServerMailer
}

export const auth = (res: HasLocals): ServerAuth => {
  const auth = res.locals.auth
  if (!auth) {
    throw new Error('No Auth object attached to server')
  }
  return auth as ServerAuth
}

export const getLocals = (res: HasLocals): Locals => {
  return {
    logger: logger(res),
    blockstore: blockstore(res),
    db: db(res),
    keypair: keypair(res),
    auth: auth(res),
    config: config(res),
    mailer: mailer(res),
  }
}
export const get = getLocals

export const plcClient = (res: Response): plc.PlcClient => {
  const cfg = config(res)
  return new plc.PlcClient(cfg.didPlcUrl)
}

export const getAuthstore = (res: Response, did: string): AuthStore => {
  const { auth, keypair } = get(res)
  // @TODO check that we can sign on behalf of this DID
  return auth.verifier.loadAuthStore(keypair, [], did)
}

export const loadRepo = async (
  res: Response,
  did: string,
): Promise<Repo | null> => {
  const { db, blockstore, auth } = getLocals(res)
  const currRoot = await db.getRepoRoot(did)
  if (!currRoot) {
    return null
  }
  const authStore = getAuthstore(res, did)
  return Repo.load(blockstore, currRoot, auth.verifier, authStore)
}
