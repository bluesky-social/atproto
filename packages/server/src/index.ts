// catch errors that get thrown in async route handlers
// this is a relatively non-invasive change to express
// they get handled in the error.handler middleware
// leave at top of file before importing Routes
import 'express-async-errors'

import express from 'express'
import cors from 'cors'
import * as auth from '@atproto/auth'
import API from './api'
import { IpldStore } from '@atproto/repo'
import Database from './db'
import ServerAuth from './auth'
import * as error from './error'
import { httpLogger, loggerMiddleware } from './logger'
import { ServerConfig, ServerConfigValues } from './config'
import { DidResolver } from '@atproto/did-resolver'
import { Locals } from './locals'
import { ServerMailer } from './mailer'
import { createTransport } from 'nodemailer'

export type { ServerConfigValues } from './config'
export { ServerConfig } from './config'
export { Database } from './db'

export type App = express.Application

const runServer = (
  blockstore: IpldStore,
  db: Database,
  keypair: auth.DidableKey,
  cfg: ServerConfigValues,
) => {
  const config = new ServerConfig(cfg)
  const didResolver = new DidResolver({ plcUrl: config.didPlcUrl })
  const auth = new ServerAuth({
    jwtSecret: cfg.jwtSecret,
    adminPass: cfg.adminPassword,
    didResolver,
  })

  const mailTransport =
    config.emailSmtpUrl !== undefined
      ? createTransport(config.emailSmtpUrl)
      : createTransport({ jsonTransport: true })

  const mailer = new ServerMailer(mailTransport, config)

  const app = express()
  app.use(cors())
  app.use(loggerMiddleware)

  const locals: Locals = {
    logger: httpLogger,
    blockstore,
    db,
    keypair,
    auth,
    config,
    mailer,
  }

  app.locals = locals

  app.use((req, res, next) => {
    const reqLocals: Locals = {
      ...locals,
      // @ts-ignore
      logger: req.log, // This logger is request-specific
    }
    res.locals = reqLocals
    next()
  })

  const apiServer = API()
  app.use(apiServer.xrpc.router)
  app.use(error.handler)

  const listener = app.listen(config.port)
  return { app, listener }
}

export default runServer
