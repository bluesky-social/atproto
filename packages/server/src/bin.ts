import dotenv from 'dotenv'
import {
  IpldStore,
  MemoryBlockstore,
  PersistentBlockstore,
} from '@atproto/repo'
import * as crypto from '@atproto/crypto'
import Database from './db'
import server from './index'
import { ServerConfig } from './config'

const run = async () => {
  const env = process.env.ENV
  if (env) {
    dotenv.config({ path: `./.${env}.env` })
  } else {
    dotenv.config()
  }

  let blockstore: IpldStore
  let db: Database

  const cfg = ServerConfig.readEnv()

  if (cfg.blockstoreLocation) {
    blockstore = new PersistentBlockstore(cfg.blockstoreLocation)
  } else {
    blockstore = new MemoryBlockstore()
  }

  if (cfg.databaseLocation) {
    db = await Database.sqlite(cfg.databaseLocation)
  } else {
    db = await Database.memory()
  }

  const keypair = await crypto.EcdsaKeypair.create()

  const { listener } = server(blockstore, db, keypair, cfg)
  listener.on('listening', () => {
    console.log(`🌞 ATP Data server is running at ${cfg.origin}`)
  })
}

run()
