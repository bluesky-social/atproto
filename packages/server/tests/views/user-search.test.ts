import AtpApi, { ServiceClient as AtpServiceClient } from '@atproto/api'
import { runTestServer, forSnapshot, CloseFn, paginateAll } from '../_util'
import { SeedClient } from '../seeds/client'
import usersBulkSeed from '../seeds/users-bulk'
import { App } from '../../src'
import * as locals from '../../src/locals'

describe('pds user search views', () => {
  let app: App
  let client: AtpServiceClient
  let close: CloseFn
  let sc: SeedClient
  let headers: { [s: string]: string }

  beforeAll(async () => {
    const server = await runTestServer({
      dbPostgresSchema: 'views_user_search',
    })
    close = server.close
    app = server.app
    client = AtpApi.service(server.url)
    sc = new SeedClient(client)
    await usersBulkSeed(sc)
    headers = sc.getHeaders(Object.values(sc.dids)[0])
  })

  afterAll(async () => {
    await close()
  })

  it('typeahead gives relevant results', async () => {
    const result = await client.app.bsky.getUsersTypeahead(
      { term: 'car' },
      undefined,
      { headers },
    )

    const names = result.data.users.map((u) => u.name)

    const shouldContain = [
      'Cara.Wiegand69',
      'Eudora_Dietrich4', // Carol Littel
      'Shane_Torphy52', // Sadie Carter
      'Aliya.Hodkiewicz', // Carlton Abernathy IV
      'Carlos6',
      'Carolina_McDermott77',
    ]

    shouldContain.forEach((name) => expect(names).toContain(name))

    if (locals.get(app).db.dialect === 'pg') {
      expect(names).toContain('Cayla_Marquardt39') // Fuzzy match supported by postgres
    } else {
      expect(names).not.toContain('Cayla_Marquardt39')
    }

    const shouldNotContain = [
      'Sven70',
      'Hilario84',
      'Santa_Hermann78',
      'Dylan61',
      'Preston_Harris',
      'Loyce95',
      'Melyna_Zboncak',
    ]

    shouldNotContain.forEach((name) => expect(names).not.toContain(name))

    if (locals.get(app).db.dialect === 'pg') {
      expect(forSnapshot(result.data.users)).toEqual(snapTypeaheadPg)
    } else {
      expect(forSnapshot(result.data.users)).toEqual(snapTypeaheadSqlite)
    }
  })

  it('typeahead gives empty result set when provided empty term', async () => {
    const result = await client.app.bsky.getUsersTypeahead(
      { term: '' },
      undefined,
      { headers },
    )

    expect(result.data.users).toEqual([])
  })

  it('typeahead applies limit', async () => {
    const full = await client.app.bsky.getUsersTypeahead(
      { term: 'p' },
      undefined,
      { headers },
    )

    expect(full.data.users.length).toBeGreaterThan(5)

    const limited = await client.app.bsky.getUsersTypeahead(
      { term: 'p', limit: 5 },
      undefined,
      { headers },
    )

    expect(limited.data.users).toEqual(full.data.users.slice(0, 5))
  })

  it('search gives relevant results', async () => {
    const result = await client.app.bsky.getUsersSearch(
      { term: 'car' },
      undefined,
      { headers },
    )

    const names = result.data.users.map((u) => u.name)

    const shouldContain = [
      'Cara.Wiegand69',
      'Eudora_Dietrich4', // Carol Littel
      'Shane_Torphy52', // Sadie Carter
      'Aliya.Hodkiewicz', // Carlton Abernathy IV
      'Carlos6',
      'Carolina_McDermott77',
    ]

    shouldContain.forEach((name) => expect(names).toContain(name))

    if (locals.get(app).db.dialect === 'pg') {
      expect(names).toContain('Cayla_Marquardt39') // Fuzzy match supported by postgres
    } else {
      expect(names).not.toContain('Cayla_Marquardt39')
    }

    const shouldNotContain = [
      'Sven70',
      'Hilario84',
      'Santa_Hermann78',
      'Dylan61',
      'Preston_Harris',
      'Loyce95',
      'Melyna_Zboncak',
    ]

    shouldNotContain.forEach((name) => expect(names).not.toContain(name))

    if (locals.get(app).db.dialect === 'pg') {
      expect(forSnapshot(result.data.users)).toEqual(snapSearchPg)
    } else {
      expect(forSnapshot(result.data.users)).toEqual(snapSearchSqlite)
    }
  })

  it('search gives empty result set when provided empty term', async () => {
    const result = await client.app.bsky.getUsersSearch(
      { term: '' },
      undefined,
      { headers },
    )

    expect(result.data.users).toEqual([])
  })

  it('paginates', async () => {
    const results = (results) => results.flatMap((res) => res.users)
    const paginator = async (cursor?: string) => {
      const res = await client.app.bsky.getUsersSearch(
        { term: 'p', before: cursor, limit: 3 },
        undefined,
        { headers },
      )
      return res.data
    }

    const paginatedAll = await paginateAll(paginator)
    paginatedAll.forEach((res) =>
      expect(res.users.length).toBeLessThanOrEqual(3),
    )

    const full = await client.app.bsky.getUsersSearch(
      { term: 'p' },
      undefined,
      { headers },
    )

    expect(full.data.users.length).toBeGreaterThan(5)
    expect(results(paginatedAll)).toEqual(results([full.data]))
  })

  it('search handles bad input', async () => {
    // Mostly for sqlite's benefit, since it uses LIKE and these are special characters that will
    // get stripped. This input triggers a special case where there are no "safe" words for sqlite to search on.
    const result = await client.app.bsky.getUsersSearch(
      { term: ' % _ ' },
      undefined,
      { headers },
    )

    expect(result.data.users).toEqual([])
  })
})

// Not using jest snapshots because it doesn't handle the conditional pg/sqlite very well:
// you can achieve it using named snapshots, but when you run the tests for pg the test suite fails
// since the sqlite snapshots appear obsolete to jest (and vice-versa when you run the sqlite suite).

const snapTypeaheadPg = [
  {
    did: 'user(0)',
    name: 'Cara.Wiegand69',
  },
  {
    did: 'user(1)',
    displayName: 'Carol Littel',
    name: 'Eudora_Dietrich4',
  },
  {
    did: 'user(2)',
    displayName: 'Sadie Carter',
    name: 'Shane_Torphy52',
  },
  {
    did: 'user(3)',
    displayName: 'Carlton Abernathy IV',
    name: 'Aliya.Hodkiewicz',
  },
  {
    did: 'user(4)',
    name: 'Carlos6',
  },
  {
    did: 'user(5)',
    displayName: 'Latoya Windler',
    name: 'Carolina_McDermott77',
  },
  {
    did: 'user(6)',
    displayName: 'Rachel Kshlerin',
    name: 'Cayla_Marquardt39',
  },
]

const snapTypeaheadSqlite = [
  {
    did: 'user(0)',
    displayName: 'Carlton Abernathy IV',
    name: 'Aliya.Hodkiewicz',
  },
  {
    did: 'user(1)',
    name: 'Cara.Wiegand69',
  },
  {
    did: 'user(2)',
    name: 'Carlos6',
  },
  {
    did: 'user(3)',
    displayName: 'Latoya Windler',
    name: 'Carolina_McDermott77',
  },
  {
    did: 'user(4)',
    displayName: 'Carol Littel',
    name: 'Eudora_Dietrich4',
  },
  {
    did: 'user(5)',
    displayName: 'Sadie Carter',
    name: 'Shane_Torphy52',
  },
]

const snapSearchPg = [
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    did: 'user(0)',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Cara.Wiegand69',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(1)',
    displayName: 'Carol Littel',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Eudora_Dietrich4',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(2)',
    displayName: 'Sadie Carter',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Shane_Torphy52',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(3)',
    displayName: 'Carlton Abernathy IV',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Aliya.Hodkiewicz',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    did: 'user(4)',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Carlos6',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(5)',
    displayName: 'Latoya Windler',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Carolina_McDermott77',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(6)',
    displayName: 'Rachel Kshlerin',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Cayla_Marquardt39',
  },
]

const snapSearchSqlite = [
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(0)',
    displayName: 'Carlton Abernathy IV',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Aliya.Hodkiewicz',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    did: 'user(1)',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Cara.Wiegand69',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    did: 'user(2)',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Carlos6',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(3)',
    displayName: 'Latoya Windler',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Carolina_McDermott77',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(4)',
    displayName: 'Carol Littel',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Eudora_Dietrich4',
  },
  {
    createdAt: '1970-01-01T00:00:00.000Z',
    description: '',
    did: 'user(5)',
    displayName: 'Sadie Carter',
    indexedAt: '1970-01-01T00:00:00.000Z',
    name: 'Shane_Torphy52',
  },
]
