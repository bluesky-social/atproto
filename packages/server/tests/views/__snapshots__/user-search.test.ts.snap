// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds user search views search gives relevant results: postgres 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.5::Cara.Wiegand69",
    "did": "user(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Cara.Wiegand69",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.57142854::Eudora_Dietrich4",
    "description": "",
    "did": "user(1)",
    "displayName": "Carol Littel",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Eudora_Dietrich4",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.625::Shane_Torphy52",
    "description": "",
    "did": "user(2)",
    "displayName": "Sadie Carter",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Shane_Torphy52",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.6666666::Aliya.Hodkiewicz",
    "description": "",
    "did": "user(3)",
    "displayName": "Carlton Abernathy IV",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Aliya.Hodkiewicz",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.6666666::Carlos6",
    "did": "user(4)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Carlos6",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.7::Carolina_McDermott77",
    "description": "",
    "did": "user(5)",
    "displayName": "Latoya Windler",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Carolina_McDermott77",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "cursor": "0.75::Cayla_Marquardt39",
    "description": "",
    "did": "user(6)",
    "displayName": "Rachel Kshlerin",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "Cayla_Marquardt39",
  },
]
`;

exports[`pds user search views typeahead gives relevant results: postgres 1`] = `
Array [
  Object {
    "did": "user(0)",
    "name": "Cara.Wiegand69",
  },
  Object {
    "did": "user(1)",
    "displayName": "Carol Littel",
    "name": "Eudora_Dietrich4",
  },
  Object {
    "did": "user(2)",
    "displayName": "Sadie Carter",
    "name": "Shane_Torphy52",
  },
  Object {
    "did": "user(3)",
    "displayName": "Carlton Abernathy IV",
    "name": "Aliya.Hodkiewicz",
  },
  Object {
    "did": "user(4)",
    "name": "Carlos6",
  },
  Object {
    "did": "user(5)",
    "displayName": "Latoya Windler",
    "name": "Carolina_McDermott77",
  },
  Object {
    "did": "user(6)",
    "displayName": "Rachel Kshlerin",
    "name": "Cayla_Marquardt39",
  },
]
`;
